### PR TITLE
Allow additional permissions for extensions in garden cluster defined via RBAC

### LINKED ISSUE #8980: Support RBAC-based cluster access for extensions
**How to categorize this issue?**

/area security usability
/kind enhancement

**What would you like to be added**:
https://github.com/gardener/gardener/issues/8001 enabled access to the `virtual-garden` cluster for extensions with almost the same permissions of Gardenlet.
Some extensions deploy further components into seed clusters that need access to the `virtual-garden` cluster, as well. For them it's desirable to use a dedicated service account token as well as customizable permissions because the ones provided for the extension are either too restrictive or too open.

Implementation proposal / approach:
1. Operators apply a `ClusterRole` in the `virtual-garden` cluster with permissions the extension's components need
2. Extensions create additional `RequestToken` secrets for components
3. Extensions create a `ClusterRoleBinding` referring to the `ClusterRole` in [1.] and the service account resulted from `[2.]`
4. The `ClusterRoleBinding` creation is protected by the `SeedRestriction` admission plugin to prevent any privilege escalation

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security usability
/kind enhancement

**What this PR does / why we need it**:
This PR adds a new controller part of `gardener-controller-manager` which watches `ClusterRole`s labeled with `authorization.extensions.gardener.cloud/additional-permissions=true`. It also watches `ServiceAccount`s in seed namespaces.

On reconciliation, the new controller creates a `ClusterRoleBinding` referencing the respective `ClusterRole`. The subjects are all `ServiceAccount`s related to extension sin seed namespaces - but only if the `ServiceAccount`s are prefixed with `extension-<name>` and `<name>` is present in the comma-separated list in the `authorization.extensions.gardener.cloud/bind-to=<name1>,<name2>` annotation.

**Which issue(s) this PR fixes**:
Fixes #8980

**Special notes for your reviewer**:
See https://github.com/gardener/gardener/issues/8980#issuecomment-1903837782

**Release note**:

```feature operator
It is now possible to define additional permissions via RBAC for extensions access in the garden cluster. You can read all about it [here](https://github.com/gardener/gardener/tree/master/docs/extensions/garden-api-access.md#additional-permissions).
```

### CODE REVIEW

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+// Reconciler reconciles ClusterRoles for additional extension permissions and creates ClusterRoleBindings for binding
+// extension service accounts to such ClusterRoles.
+type Reconciler struct {
+	Client client.Client
+}
+
+// Reconcile reconciles ControllerRegistrations.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```
Comment (timuthy) [inline pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:43]: I don't see a clean-up logic for `ClusterRoleBinding`s, e.g. when the seed selector of the `ControllerRegistration` is used. Souldn't we have this?

File: docs/extensions/garden-api-access.md [hunk 2]
```diff
@@ -127,8 +127,36 @@ With this, extensions are restricted to work with objects in the garden cluster
 Note that if the plugins are not enabled, extension clients are only granted read access to global resources like `CloudProfiles` (this is granted to all authenticated users).
 There are a few exceptions to the granted permissions as documented [here](../deployment/gardenlet_api_access.md#rule-exceptions-for-extension-clients).
 
+### Additional Permissions
```
Comment (timuthy) [inline docs/extensions/garden-api-access.md:130]: I don't think that this was the exact intention of the issue which says:

> use a dedicated service account token as well as customizable permissions because the ones provided for the extension are either too restrictive or too open.

Additional permissions solve the "too restrictive" case but doesn't cover the "too open" one. While being in clarification with @nickytd, I understood that the goal is not to fully inherit existing permissions to the new `ServiceAccount`. For security reasons, it should only have the ones specific in the `ClusterRole`.

Is this already supported? IIRC, the Seed Authorizer relies on a `extension` prefix ...

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		if err := r.Client.List(ctx, serviceAccountList, client.InNamespace(namespace.Name)); err != nil {
+			return nil, fmt.Errorf("failed listing ServiceAccounts in namespace %s: %w", namespace.Name, err)
+		}
+
+		// ensure stable order of subjects
+		kubernetesutils.ByName().Sort(serviceAccountList)
+
+		for _, serviceAccount := range serviceAccountList.Items {
+			for _, extensionName := range extensionNames {
+				if strings.HasPrefix(serviceAccount.Name, v1beta1constants.ExtensionGardenServiceAccountPrefix+extensionName) {
```
Comment (timuthy) [inline pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:114]: I noticed that matching the `extensionName` has some drawbacks:
- An extension cannot create multiple `ServiceAccount`s with distinct permissions, since the permissions are aggregated.
- We calculate `ClusterRoleBinding`s again and again due to the `ServiceAccount` watch, i.e. it has no direct relation to the desired `ClusterRole`.

How about if we don't use the extension but the service account name for matching?

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: extension-create-subjectaccessreviews
  annotations:
    authorization.extensions.gardener.cloud/bind-to: <service-account-name>
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks for the implementation 👍
I have some questions and doubts after the review, though - please see the comments for more information.

File: docs/extensions/garden-api-access.md (hunk 2 above)
Comment (rfranzke) [reply docs/extensions/garden-api-access.md:130]: Reducing the permissions is currently not possible - this PR is only about extending them. I'm also unsure if the current approach would be a good way to implement it (I didn't think about this use case yet, but I guess it will not work well together with the seed authorizer (e.g., if the `ClusterRole` allowed to delete shoots, then the extension could just delete all shoots, no matter whether they relate to its seed or not, since RBAC would already allow the request (and seed authorizer won't be asked about it))). This is probably not what we want to achieve.

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go (hunk 1 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:43]: Can you make your example more concrete? I assume you talk about a `ControllerRegistration` being changed with a seed selector that leads to deinstallation of an extension from a certain seed. In this case, the `ServiceAccount` will be deleted which triggers a new reconciliation (and with this, a re-computation of the subjects for the `ClusterRoleBinding`).

Comment (timuthy) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:43]: You're right, thanks for clarifying.

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go (hunk 3 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:114]: The problem with using the name of the service account is that it is not static/known upfront. `gardenlet` uses the extension namespace name in its seed for the service account name, and this is different in each seed/for each installation.

TBH, I don't have good answers for the points above (especially the first one). I guess this is somewhat related to the concept of reducing permissions (which we don't have today).

File: docs/extensions/garden-api-access.md (hunk 2 above)
Comment (timuthy) [reply docs/extensions/garden-api-access.md:130]: That's already the case now, isn't it?
The following `ClusterRole` will allow the `ServiceAccount`(s) to delete random shoots with the current implementation:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    authorization.extensions.gardener.cloud/bind-to: provider-local
  labels:
    authorization.extensions.gardener.cloud/additional-permissions: "true"
  name: extension-shoot-mutation
rules:
- apiGroups:
  - core.gardener.cloud
  resources:
  - shoots
  verbs:
  - '*'
```

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go (hunk 3 above)
Comment (timuthy) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:114]: > The problem with using the name of the service account is that it is not static/known upfront. gardenlet uses the extension namespace name in its seed for the service account name, and this is different in each seed/for each installation.

Doesn't that only happen for the automatically created accounts? The extensions can choose any name they want when requesting another service account, also a static one ...

Comment (rfranzke) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:114]: Sure, but this also means that any reasonable usage of this feature requires to create new `ServiceAccount`s just because the name needs to be static. I'm not sure whether that's desireable.

File: docs/extensions/garden-api-access.md (hunk 2 above)
Comment (rfranzke) [reply docs/extensions/garden-api-access.md:130]: True :) All these downsides are reasons for not having invested into these features (changing extension permissions) earlier. I also doubt that we can find a perfect solution just by using RBAC (otherwise, we wouldn't have implemented the seed authorizer for the gardenlet permissions).
FMPOV, extending the permissions as proposed in this PR is acceptable (with the risk of human operators to make mistake like [this](https://github.com/gardener/gardener/pull/9066#discussion_r1464730794) (though, this is already the case today, i.e., human operators can easily grant too wide permissions via RBAC which bypass the seed authorizer)). Reducing the permissions is more complicated and I would continue to refrain from looking into it for now (especially in the light that all extensions run in the same seed and can easily "steal" the gardenlet kubeconfig, so the overall security improvement is rather limited/minimal).

An alternative to this PR is to do what you apparently originally had in mind, i.e., allow human operators to define a `ClusterRole` for the extensions and let them completely bypass the seed authorizer. TBH, I'm not too much in favor of a concept like this which relies on "intentionally deactivating the seed authorizer" (while [this](https://github.com/gardener/gardener/pull/9066#discussion_r1464730794) rather looks like a mistake/error).
I understand that this would be sufficient/suitable for @nickytd's use case, but that's probably a pretty specific one. My expectation would be that a more common scenario is to add additional permissions to extensions for resources not covered by the seed authorizer (e.g., the `SubjectAccessReview`s).

TL;DR: I'm also not fully convinced by this PR, but I think that it is a good compromise without being a solution that's "too special" for this single scenario.

File: pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go (hunk 3 above)
Comment (timuthy) [reply pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go:114]: > Sure, but this also means that any reasonable usage of this feature requires to create new ServiceAccounts just because the name needs to be static. I'm not sure whether that's desireable.

That was exactly the reason for #8980: Extensions request a separate and dedicated service account for a component that is additionally deployed by them to the seed. It was actually never intended to re-use the existing one.

File: docs/extensions/garden-api-access.md (hunk 2 above)
Comment (timuthy) [reply docs/extensions/garden-api-access.md:130]: >  Reducing the permissions is more complicated and I would continue to refrain from looking into it for now

Is it really that complex? If we choose to follow the name matching (https://github.com/gardener/gardener/pull/9066#discussion_r1464598490), it'll come out of the box.

**Comment by rfranzke** [comment]
I'm closing this PR in the light of https://github.com/gardener/gardener/issues/8980#issuecomment-1908051623 to simplify tracking of the required changes described in the issue. I'll work on those and open a new PR 👍🏻 

/close
